algorithms/src/std_algorithms/Kokkos_HelperPredicates.hpp

```diff
@@ -0,0 +1,115 @@
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_STD_HELPER_PREDICATES_HPP
+#define KOKKOS_STD_HELPER_PREDICATES_HPP
+
+// naming convetion:
+// StdAlgoSomeExpressiveNameUnaryPredicate
+// StdAlgoSomeExpressiveNameBinaryPredicate
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+
+// ------------------
+// UNARY PREDICATES
+// ------------------
+template <class T>
+struct StdAlgoEqualsValUnaryPredicate {
+  T m_value;
+
+  KOKKOS_INLINE_FUNCTION
+  bool operator()(const T& val) const { return val == m_value; }
+
+  KOKKOS_INLINE_FUNCTION
+  StdAlgoEqualsValUnaryPredicate(const T& _value) : m_value(_value) {}
+};
+
+template <class T>
+struct StdAlgoNotEqualsValUnaryPredicate {
+  T m_value;
+
+  KOKKOS_INLINE_FUNCTION
+  bool operator()(const T& val) const { return !(val == m_value); }
+
+  KOKKOS_INLINE_FUNCTION
+  StdAlgoNotEqualsValUnaryPredicate(const T& _value) : m_value(_value) {}
+};
+
+template <class ValueType, class PredicateType>
+struct StdAlgoNegateUnaryPredicateWrapper {
+  PredicateType m_pred;
+
+  KOKKOS_INLINE_FUNCTION
+  bool operator()(const ValueType& val) const { return !m_pred(val); }
+
+  KOKKOS_INLINE_FUNCTION
+  StdAlgoNegateUnaryPredicateWrapper(const PredicateType& pred)
+      : m_pred(pred) {}
+};
+
+// ------------------
+// BINARY PREDICATES
+// ------------------
+template <class ValueType1, class ValueType2>
```

Comment:
Should these be:

`template<class ValueType1, class ValueType2 = ValueType1>`
or even

```
template<class ValueType1 = void, class ValueType2 = ValueType1>
struct StdAlgoEqualBinaryPredicate {
   KOKKOS_INLINE_FUNCTION
   bool operator()(const ValueType1& a, const ValueType2& b) const {
     return a == b;
   }
 };

template<>
struct StdAlgoEqualBinaryPredicate<void> {
   KOKKOS_INLINE_FUNCTION
   template<typename ValueType1, typename ValueType2>
   bool operator()(const ValueType1& a, const ValueType2& b) const
   { return a == b; }

   using is_transparent = `unspecified`; // could probably be true_type
};
```

Reply:
Yes, I agree on this `template<class ValueType1, class ValueType2>` . 

Not sure I grasp the advantage of the second solution ?

Reply:
I changed it to use a default as suggested. Do we leave that way or do we want to use the `is_transparent` version?